flang/lib/Lower/IntrinsicCall.cpp

```diff
@@ -2136,6 +2142,86 @@ IntrinsicLibrary::genSum(mlir::Type resultType,
                       "unexpected result for Sum", args);
 }
 
+// TRANSFER
+fir::ExtendedValue
+IntrinsicLibrary::genTransfer(mlir::Type resultType,
+                              llvm::ArrayRef<fir::ExtendedValue> args) {
+
+  assert(args.size() >= 2); // args.size() == 2 when size argument is omitted.
+
+  // Handle source argument
+  auto source = builder.createBox(loc, args[0]);
+
+  // Handle mold argument
+  auto mold = builder.createBox(loc, args[1]);
+  fir::BoxValue moldTmp = builder.createBox(loc, args[1]);
+  auto moldRank = moldTmp.rank();
+
+  bool absentSize = (args.size() == 2);
+
+  if (moldRank == 0 && absentSize) {
```

Comment:
I think you could merge both codes with:
```
auto type = (moldRank == 0 && absentSize)  ? resultType : builder.getVarLenSeqTy(resultType, 1);
auto resultMutableBox =
        Fortran::lower::createTempMutableBox(builder, loc, type);
```
and merging the two res.match() into a single one.